### DISCUSSION TITLE
Welcome to rust-gpu Discussions!

### DISCUSSION THREAD

**Discussion by XAMPPRocky** [EmbarkStudios/rust-gpu#491] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Share shaders you've made.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by experiment9123** [comment]
intro - I'm a former console gamedev (20 years low level/graphics programming on a range of platfoms) dabbling with various things, I've been into Rust on and off for 5+ years. I've done some simple GL rendering experiments in rust a few years ago and am looking at jumping back in .

Question - could this 'rust-gpu' project handle compute shaders and do something like [SYCL](https://developer.codeplay.com/products/computecpp/ce/guides/sycl-guide/hello-sycl) (embed GPU compute kernel code directly inside your main program - 'single source approach' to GPU programming). One important capability would be generating compute kernels using generics. (at this minute i'm brainstorming some ideas with spiking neuron people). 

If i've understood right I appear to be seeing this isgenerating SPIR-V from rust sources, and SPIR-V unifies graphics & compute, right?

**Reply by repi** [reply]
Welcome fellow old console gamedev :) Sorry about the late reply, didn't see until now.

> Question - could this 'rust-gpu' project handle compute shaders and do something like SYCL (embed GPU compute kernel code directly inside your main program - 'single source approach' to GPU programming).

Compiling both CPU code and GPU code from single source is not something that is possible today, and looks hard to do with Cargo in general. But what one can do, and what we do for our shaders, is to share code and crates and compile them separately both for CPU and GPU.

> If i've understood right I appear to be seeing this isgenerating SPIR-V from rust sources, and SPIR-V unifies graphics & compute, right?

Yup this is a Rust compiler backend that takes the Rust intermediate representation and compiles it directly to SPIR-V, and works for both graphics and compute shaders.

**Comment by zicklag** [comment]
Hey there! I'm an experienced Rust programmer that is just getting into rendering and graphics, with just a little bit of super-simple 2D OpenGL rendering under my belt.

I'm super excited about Rust GPU! Being able to write GPU code in Rust, could have an incredible effect on the GPU ecosystem.

In the compute field, too, combined with WGPU it seems like we might finally have an option for GPU accelerated compute that works on all major platforms, and eventually even the browser. We could make machine learning libraries that run on CPU, Vulkan, Metal, DirectX, and OpenGL, all with the same code! No more CUDA NVidia lock-in! We'll see what happens, but this could be huge. :smiley:

**Reply by repi** [reply]
Glad to hear it! We are really excited about the possibilities as well. 

Our main focus is graphics & compute for graphics through Vulkan, but would love to see that extend to WebGPU and the browser in the future. And in the long term future general GPU compute in Rust instead of OpenCL/CUDA would be fantastic, but a lot of extra works and would require a quite large and active community to build and support it.
